geomstats/geometry/complex_poincare_disk.py

```diff
@@ -0,0 +1,360 @@
+        -------
+        belongs : array-like, shape=[...,]
+            Boolean denoting if point belongs to the
+            complex Poincaré disk.
+        """
+        is_scalar = gs.ndim(point) == 1
+        is_scalar = is_scalar or (gs.ndim(point) == 2 and point.shape[1] == 1)
+        has_modulus_lower_than_one = gs.array(
+            [[gs.all(gs.abs(point[i_point]) < 1)] for i_point in range(point.shape[0])]
+        )
+        belongs = gs.logical_and(is_scalar, has_modulus_lower_than_one)
+        belongs = gs.reshape(belongs, (-1,))
+        return belongs
+
+    @staticmethod
+    def projection(point):
+        """Project a point on the complex unit disk.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., 1]
+            Point to project.
+        atol : float
+            Tolerance.
+            Optional, default: backend atol.
+
+        Returns
+        -------
+        projected : array-like, shape=[..., 1]
+            Projected point.
+        """
+        return gs.where(
+            gs.abs(point) >= 1 - gs.atol,
+            gs.cast((1 - gs.atol) / gs.abs(point), dtype=point.dtype) * point,
+            point,
+        )
+
+    @staticmethod
+    def random_point(n_samples=1, bound=0.8):
+        """Generate random points in the complex unit disk.
+
+        Parameters
+        ----------
+        n_samples : int
+            Number of samples.
+            Optional, default: 1.
+        bound : float
+            Bound of the interval in which to sample in the tangent space.
+            Optional, default: 0.8.
+
+        Returns
+        -------
+        samples : array-like, shape=[..., 1]
+            Points sampled in the unit disk.
+        """
+        size = (n_samples, 1) if n_samples != 1 else (1,)
+
+        modulus = gs.random.rand(*size, dtype=gs.get_default_cdtype())
+        modulus *= bound
+        angle = 2 * gs.pi * gs.random.rand(*size, dtype=gs.get_default_cdtype())
+        samples = modulus * gs.exp(1j * angle)
+        return samples
+
+
+class ComplexPoincareDiskMetric(ComplexRiemannianMetric):
+    """Class for the complex Poincaré metric.
+
+    Parameters
+    ----------
+    scale : float
+        Scale of the complex Poincaré metric.
+        Optional, default: 1.
+    """
+
+    def __init__(self, scale=1.0):
+        super().__init__(
+            dim=1,
+            shape=(1,),
+            signature=(1, 0),
+        )
+        self.scale = scale
+
+    def inner_product_matrix(self, base_point):
+        """Compute the inner product matrix at base point.
+
+        Parameters
+        ----------
+        base_point : array-like, shape=[..., 1]
+            Base point.
+
+        Returns
+        -------
+        inner_prod_mat : array-like, shape=[..., 1]
+            Inner product matrix.
+        """
+        inner_prod_mat = 1 / (1 - gs.abs(base_point) ** 2) ** 2
+        inner_prod_mat = gs.cast(inner_prod_mat, dtype=gs.get_default_cdtype())
```

Comment:
it seems unnecessary to cast.

Reply:
If I do not cast, TensorFlow gives me the following error:

```
========================================================================================================== FAILURES ===========================================================================================================
__________________________________________________________ TestComplexPoincareDiskMetric.test_inner_product[0.5-tangent_vec_a0-tangent_vec_b0-base_point0-expected0] __________________________________________________________

self = <tests.tests_geomstats.test_complex_poincare_disk.TestComplexPoincareDiskMetric object at 0x7f97b6c098b0>, scale = 0.5, tangent_vec_a = [1j], tangent_vec_b = [1j], base_point = [0j], expected = [0.25]

    def test_inner_product(
        self, scale, tangent_vec_a, tangent_vec_b, base_point, expected
    ):
        metric = self.Metric(scale)
>       result = metric.inner_product(
            gs.array(tangent_vec_a),
            gs.array(tangent_vec_b),
            gs.array(base_point),
        )

tests/tests_geomstats/test_complex_poincare_disk.py:57: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
geomstats/geometry/complex_poincare_disk.py:180: in inner_product
    inner_prod = gs.conj(tangent_vec_a) * inner_product_matrix * tangent_vec_b
venv/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py:141: in error_handler
    return fn(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:1407: in binary_op_wrapper
    return func(x, y, name=name)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:1767: in _mul_dispatch
    return multiply(x, y, name=name)
venv/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py:141: in error_handler
    return fn(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py:1176: in op_dispatch_handler
    return dispatch_target(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:529: in multiply
    return gen_math_ops.mul(x, y, name)
venv/lib/python3.8/site-packages/tensorflow/python/ops/gen_math_ops.py:6581: in mul
    _ops.raise_from_not_ok_status(e, name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

e = _NotOkStatusException(), name = None

    def raise_from_not_ok_status(e, name):
      e.message += (" name: " + name if name is not None else "")
>     raise core._status_to_exception(e) from None  # pylint: disable=protected-access
E     tensorflow.python.framework.errors_impl.InvalidArgumentError: cannot compute Mul as input #1(zero-based) was expected to be a complex128 tensor but is a double tensor [Op:Mul]

venv/lib/python3.8/site-packages/tensorflow/python/framework/ops.py:7215: InvalidArgumentError
___________________________________________________________ TestComplexPoincareDiskMetric.test_inner_product[1-tangent_vec_a1-tangent_vec_b1-base_point1-expected1] ___________________________________________________________

self = <tests.tests_geomstats.test_complex_poincare_disk.TestComplexPoincareDiskMetric object at 0x7f97b6c09940>, scale = 1, tangent_vec_a = [(3+4j)], tangent_vec_b = [(3+4j)], base_point = [0.0], expected = [25]

    def test_inner_product(
        self, scale, tangent_vec_a, tangent_vec_b, base_point, expected
    ):
        metric = self.Metric(scale)
>       result = metric.inner_product(
            gs.array(tangent_vec_a),
            gs.array(tangent_vec_b),
            gs.array(base_point),
        )

tests/tests_geomstats/test_complex_poincare_disk.py:57: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
geomstats/geometry/complex_poincare_disk.py:180: in inner_product
    inner_prod = gs.conj(tangent_vec_a) * inner_product_matrix * tangent_vec_b
venv/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py:141: in error_handler
    return fn(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:1407: in binary_op_wrapper
    return func(x, y, name=name)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:1767: in _mul_dispatch
    return multiply(x, y, name=name)
venv/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py:141: in error_handler
    return fn(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py:1176: in op_dispatch_handler
    return dispatch_target(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:529: in multiply
    return gen_math_ops.mul(x, y, name)
venv/lib/python3.8/site-packages/tensorflow/python/ops/gen_math_ops.py:6581: in mul
    _ops.raise_from_not_ok_status(e, name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

e = _NotOkStatusException(), name = None

    def raise_from_not_ok_status(e, name):
      e.message += (" name: " + name if name is not None else "")
>     raise core._status_to_exception(e) from None  # pylint: disable=protected-access
E     tensorflow.python.framework.errors_impl.InvalidArgumentError: cannot compute Mul as input #1(zero-based) was expected to be a complex128 tensor but is a double tensor [Op:Mul]

venv/lib/python3.8/site-packages/tensorflow/python/framework/ops.py:7215: InvalidArgumentError
___________________________________________________________ TestComplexPoincareDiskMetric.test_inner_product[2-tangent_vec_a2-tangent_vec_b2-base_point2-expected2] ___________________________________________________________

self = <tests.tests_geomstats.test_complex_poincare_disk.TestComplexPoincareDiskMetric object at 0x7f97b6c099d0>, scale = 2, tangent_vec_a = [1j, 1j, 1j], tangent_vec_b = [1j, 2j, 3j], base_point = [0j, 0j, 0j]
expected = [4, 8, 12]

    def test_inner_product(
        self, scale, tangent_vec_a, tangent_vec_b, base_point, expected
    ):
        metric = self.Metric(scale)
>       result = metric.inner_product(
            gs.array(tangent_vec_a),
            gs.array(tangent_vec_b),
            gs.array(base_point),
        )

tests/tests_geomstats/test_complex_poincare_disk.py:57: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
geomstats/geometry/complex_poincare_disk.py:180: in inner_product
    inner_prod = gs.conj(tangent_vec_a) * inner_product_matrix * tangent_vec_b
venv/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py:141: in error_handler
    return fn(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:1407: in binary_op_wrapper
    return func(x, y, name=name)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:1767: in _mul_dispatch
    return multiply(x, y, name=name)
venv/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py:141: in error_handler
    return fn(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py:1176: in op_dispatch_handler
    return dispatch_target(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:529: in multiply
    return gen_math_ops.mul(x, y, name)
venv/lib/python3.8/site-packages/tensorflow/python/ops/gen_math_ops.py:6581: in mul
    _ops.raise_from_not_ok_status(e, name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

e = _NotOkStatusException(), name = None

    def raise_from_not_ok_status(e, name):
      e.message += (" name: " + name if name is not None else "")
>     raise core._status_to_exception(e) from None  # pylint: disable=protected-access
E     tensorflow.python.framework.errors_impl.InvalidArgumentError: cannot compute Mul as input #1(zero-based) was expected to be a complex128 tensor but is a double tensor [Op:Mul]

venv/lib/python3.8/site-packages/tensorflow/python/framework/ops.py:7215: InvalidArgumentError
_____________________________________________________________ TestComplexPoincareDiskMetric.test_dist_is_norm_of_log[metric_args0-point_a0-point_b0-1e-05-1e-06] ______________________________________________________________

self = <tests.tests_geomstats.test_complex_poincare_disk.TestComplexPoincareDiskMetric object at 0x7f97b6c156a0>, metric_args = (4,)
point_a = <tf.Tensor: shape=(1,), dtype=complex128, numpy=array([0.00143344+0.00245192j])>, point_b = <tf.Tensor: shape=(1,), dtype=complex128, numpy=array([-0.0026443-0.0037707j])>, rtol = 1e-05, atol = 1e-06

    def test_dist_is_norm_of_log(self, metric_args, point_a, point_b, rtol, atol):
        """Check that distance is norm of log.
    
        Parameters
        ----------
        metric_args : tuple
            Arguments to pass to constructor of the metric.
        point_a : array-like
            Point on the manifold.
        point_b : array-like
            Point on the manifold.
        rtol : float
            Relative tolerance to test this property.
        atol : float
            Absolute tolerance to test this property.
        """
        metric = self.Metric(*metric_args)
>       log = metric.norm(metric.log(point_a, point_b), point_b)

tests/geometry_test_cases.py:1193: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
geomstats/geometry/riemannian_metric.py:267: in norm
    sq_norm = self.squared_norm(vector, base_point)
geomstats/geometry/complex_poincare_disk.py:202: in squared_norm
    sq_norm = self.inner_product(vector, vector, base_point=base_point)
geomstats/geometry/complex_poincare_disk.py:180: in inner_product
    inner_prod = gs.conj(tangent_vec_a) * inner_product_matrix * tangent_vec_b
venv/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py:141: in error_handler
    return fn(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:1407: in binary_op_wrapper
    return func(x, y, name=name)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:1767: in _mul_dispatch
    return multiply(x, y, name=name)
venv/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py:141: in error_handler
    return fn(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py:1176: in op_dispatch_handler
    return dispatch_target(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:529: in multiply
    return gen_math_ops.mul(x, y, name)
venv/lib/python3.8/site-packages/tensorflow/python/ops/gen_math_ops.py:6581: in mul
    _ops.raise_from_not_ok_status(e, name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

e = _NotOkStatusException(), name = None

    def raise_from_not_ok_status(e, name):
      e.message += (" name: " + name if name is not None else "")
>     raise core._status_to_exception(e) from None  # pylint: disable=protected-access
E     tensorflow.python.framework.errors_impl.InvalidArgumentError: cannot compute Mul as input #1(zero-based) was expected to be a complex128 tensor but is a double tensor [Op:Mul]

venv/lib/python3.8/site-packages/tensorflow/python/framework/ops.py:7215: InvalidArgumentError
______________________________________________ TestComplexPoincareDiskMetric.test_inner_product_is_hermitian[metric_args0-tangent_vec_a0-tangent_vec_b0-base_point0-1e-05-1e-06] ______________________________________________

self = <tests.tests_geomstats.test_complex_poincare_disk.TestComplexPoincareDiskMetric object at 0x7f97b6c257f0>, metric_args = (4,)
tangent_vec_a = <tf.Tensor: shape=(1, 1), dtype=complex128, numpy=array([[0.2402022+1.79025119j]])>, tangent_vec_b = <tf.Tensor: shape=(1, 1), dtype=complex128, numpy=array([[0.52188672+0.74501778j]])>
base_point = <tf.Tensor: shape=(1,), dtype=complex128, numpy=array([-0.02215659-0.3258108j])>, rtol = 1e-05, atol = 1e-06

    def test_inner_product_is_hermitian(
        self, metric_args, tangent_vec_a, tangent_vec_b, base_point, rtol, atol
    ):
        """Check that the inner product is Hermitian.
    
        Parameters
        ----------
        metric_args : tuple
            Arguments to pass to constructor of the metric.
        tangent_vec_a : array-like
            Tangent vector to the manifold at base_point.
        tangent_vec_b : array-like
            Tangent vector to the manifold at base_point.
        base_point : array-like
            Point on manifold.
        rtol : float
            Relative tolerance to test this property.
        atol : float
            Absolute tolerance to test this property.
        """
        metric = self.Metric(*metric_args)
>       ip_a_b = metric.inner_product(tangent_vec_a, tangent_vec_b, base_point)

tests/geometry_test_cases.py:1363: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
geomstats/geometry/complex_poincare_disk.py:180: in inner_product
    inner_prod = gs.conj(tangent_vec_a) * inner_product_matrix * tangent_vec_b
venv/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py:141: in error_handler
    return fn(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:1407: in binary_op_wrapper
    return func(x, y, name=name)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:1767: in _mul_dispatch
    return multiply(x, y, name=name)
venv/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py:141: in error_handler
    return fn(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py:1176: in op_dispatch_handler
    return dispatch_target(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:529: in multiply
    return gen_math_ops.mul(x, y, name)
venv/lib/python3.8/site-packages/tensorflow/python/ops/gen_math_ops.py:6581: in mul
    _ops.raise_from_not_ok_status(e, name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

e = _NotOkStatusException(), name = None

    def raise_from_not_ok_status(e, name):
      e.message += (" name: " + name if name is not None else "")
>     raise core._status_to_exception(e) from None  # pylint: disable=protected-access
E     tensorflow.python.framework.errors_impl.InvalidArgumentError: cannot compute Mul as input #1(zero-based) was expected to be a complex128 tensor but is a double tensor [Op:Mul]

venv/lib/python3.8/site-packages/tensorflow/python/framework/ops.py:7215: InvalidArgumentError
______________________________________________ TestComplexPoincareDiskMetric.test_inner_product_is_hermitian[metric_args1-tangent_vec_a1-tangent_vec_b1-base_point1-1e-05-1e-06] ______________________________________________

self = <tests.tests_geomstats.test_complex_poincare_disk.TestComplexPoincareDiskMetric object at 0x7f97b6c25880>, metric_args = (3,)
tangent_vec_a = <tf.Tensor: shape=(2, 1), dtype=complex128, numpy=
array([[-0.71589057-0.17066383j],
       [ 0.72516403-1.93261272j]])>
tangent_vec_b = <tf.Tensor: shape=(2, 1), dtype=complex128, numpy=
array([[0.2385537 +0.65787477j],
       [0.79146253+0.84277774j]])>
base_point = <tf.Tensor: shape=(1,), dtype=complex128, numpy=array([-0.19719644-0.54558861j])>, rtol = 1e-05, atol = 1e-06

    def test_inner_product_is_hermitian(
        self, metric_args, tangent_vec_a, tangent_vec_b, base_point, rtol, atol
    ):
        """Check that the inner product is Hermitian.
    
        Parameters
        ----------
        metric_args : tuple
            Arguments to pass to constructor of the metric.
        tangent_vec_a : array-like
            Tangent vector to the manifold at base_point.
        tangent_vec_b : array-like
            Tangent vector to the manifold at base_point.
        base_point : array-like
            Point on manifold.
        rtol : float
            Relative tolerance to test this property.
        atol : float
            Absolute tolerance to test this property.
        """
        metric = self.Metric(*metric_args)
>       ip_a_b = metric.inner_product(tangent_vec_a, tangent_vec_b, base_point)

tests/geometry_test_cases.py:1363: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
geomstats/geometry/complex_poincare_disk.py:180: in inner_product
    inner_prod = gs.conj(tangent_vec_a) * inner_product_matrix * tangent_vec_b
venv/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py:141: in error_handler
    return fn(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:1407: in binary_op_wrapper
    return func(x, y, name=name)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:1767: in _mul_dispatch
    return multiply(x, y, name=name)
venv/lib/python3.8/site-packages/tensorflow/python/util/traceback_utils.py:141: in error_handler
    return fn(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py:1176: in op_dispatch_handler
    return dispatch_target(*args, **kwargs)
venv/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py:529: in multiply
    return gen_math_ops.mul(x, y, name)
venv/lib/python3.8/site-packages/tensorflow/python/ops/gen_math_ops.py:6581: in mul
    _ops.raise_from_not_ok_status(e, name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

e = _NotOkStatusException(), name = None

    def raise_from_not_ok_status(e, name):
      e.message += (" name: " + name if name is not None else "")
>     raise core._status_to_exception(e) from None  # pylint: disable=protected-access
E     tensorflow.python.framework.errors_impl.InvalidArgumentError: cannot compute Mul as input #1(zero-based) was expected to be a complex128 tensor but is a double tensor [Op:Mul]

venv/lib/python3.8/site-packages/tensorflow/python/framework/ops.py:7215: InvalidArgumentError
====================================================================================================== warnings summary =======================================================================================================
venv/lib/python3.8/site-packages/flatbuffers/compat.py:19
  /home/ycabanes/work/geomstats/geomstats/venv/lib/python3.8/site-packages/flatbuffers/compat.py:19: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

venv/lib/python3.8/site-packages/tensorflow_probability/python/__init__.py:57
  /home/ycabanes/work/geomstats/geomstats/venv/lib/python3.8/site-packages/tensorflow_probability/python/__init__.py:57: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    if (distutils.version.LooseVersion(tf.__version__) <

venv/lib/python3.8/site-packages/tensorflow_probability/python/__init__.py:58
  /home/ycabanes/work/geomstats/geomstats/venv/lib/python3.8/site-packages/tensorflow_probability/python/__init__.py:58: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    distutils.version.LooseVersion(required_tensorflow_version)):

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=================================================================================================== short test summary info ===================================================================================================
FAILED tests/tests_geomstats/test_complex_poincare_disk.py::TestComplexPoincareDiskMetric::test_inner_product[0.5-tangent_vec_a0-tangent_vec_b0-base_point0-expected0] - tensorflow.python.framework.errors_impl.InvalidArgu...
FAILED tests/tests_geomstats/test_complex_poincare_disk.py::TestComplexPoincareDiskMetric::test_inner_product[1-tangent_vec_a1-tangent_vec_b1-base_point1-expected1] - tensorflow.python.framework.errors_impl.InvalidArgume...
FAILED tests/tests_geomstats/test_complex_poincare_disk.py::TestComplexPoincareDiskMetric::test_inner_product[2-tangent_vec_a2-tangent_vec_b2-base_point2-expected2] - tensorflow.python.framework.errors_impl.InvalidArgume...
FAILED tests/tests_geomstats/test_complex_poincare_disk.py::TestComplexPoincareDiskMetric::test_dist_is_norm_of_log[metric_args0-point_a0-point_b0-1e-05-1e-06] - tensorflow.python.framework.errors_impl.InvalidArgumentErr...
FAILED tests/tests_geomstats/test_complex_poincare_disk.py::TestComplexPoincareDiskMetric::test_inner_product_is_hermitian[metric_args0-tangent_vec_a0-tangent_vec_b0-base_point0-1e-05-1e-06] - tensorflow.python.framework...
FAILED tests/tests_geomstats/test_complex_poincare_disk.py::TestComplexPoincareDiskMetric::test_inner_product_is_hermitian[metric_args1-tangent_vec_a1-tangent_vec_b1-base_point1-1e-05-1e-06] - tensorflow.python.framework...
==================================================================================== 6 failed, 37 passed, 19 skipped, 3 warnings in 2.19s =====================================================================================
```